### ISSUE TITLE
Api for chatrooms

### ISSUE THREAD

**Issue by seshubonam** [dappros/ethora#498]
**Is your feature request related to a problem? Please describe.**
Need a way to connect to chatroom messages to read and write messages and media. Also need to get usernames from the chatroom to show recent interactions outside the chat window, in the home screen 

**Describe the solution you'd like**
For a voice chat app that takes user posted voice notes and converts to ai generated videos with avatars/scenes related to the voice note. this app enables a creative way of sharing opinions with friends and followers. 
Here is our requirement:
1. we have room list which shows tiles of latest messages from each room. we need a way to access each room latest messages/media posted to a specific room and the username of the member posting. 
2. we will be recording voice notes and media outside the chat window using a bottom sheet view that opens upon clicking a room name. this media after being created should be posted to the chatroom(the specific chatroom that user clicks to open its respective bottom sheet view)

**Describe alternatives you've considered**
Alternative would be to increase the number of steps to send a voice-note/media by directing the user to open chat view and then selecting options to record voice or attach media by modifying these inputs to fir our custom needs

**Additional context**
attaching a video of the app development done so far. 

in the attached video, the user flow is as follows:
1. the home screen shows a list of rooms with tiles. each tile is a voice note or a video note posted to the respective room. 2. upon clicking any tile a story component like instagram stories is played, which houses the voice-note/video(media) posted by a member of the room.
3. users can click the "mic" button on left side of each room in home screen to post media to the room.
4. for further text chat, a "message" icon can be clicked which directs user to the regular ethora chat room with a slightly modified ui that shows the story tiles at the top of the screen. 

In this process, API to chat room messages/media and usernames with read write access is required to fill the home screen tiles with relevant information about the latest media updates from the chat room.

what is the best way we could proceed. 

demo video on ios simulator: https://github.com/dappros/ethora/assets/1757478/929581be-9147-4c35-92c2-92896947033a

please let me know if my description is clear.

**Comment by dzinzyura** [comment]
Hello, @seshubonam !

Thank you for your question and your interest for the Ethora project! 

Based on my understanding, it seems you're seeking a comprehensive API description. To access all the necessary details for implementing the desired features, you can follow this link: 

http://api.ethoradev.com/api-docs 

and here's the link for the XMPP documentation as well: 

https://www.npmjs.com/package/@xmpp/client (here you can find the examples of connection)
https://xmpp.org/extensions/xep-0045.html (here - is how to work with rooms)

Should you have any additional inquiries, please don't hesitate to reach out. We're here to assist you.

Best regards,
Ethora team

*Feel free to reopen this issue if you have any further questions.

**Comment by seshubonam** [comment]
Thank you, will check and follow the docs.

**Comment by phwizard** [comment]
@seshubonam 
I'm not really sure what's the best way to achieve that but certainly there are options. Thoughts below:

**Rooms and messages API** - our REST JSON Rooms API is mainly to manage the rooms but for the messages we normally rely on XMPP protocol. Let's clarify the requirement here and we can help. Do you want to simply be able to read N latest messages from each chat room? Or do you also want to post messages into room using REST JSON API? 

**Latest messages** - in our engine the last message from each chat room is already retrieved by default for the rooms list view. We could modify that mechanism to include more messages. How many latest messages do you think you need? Do you only want to include messages of a specific type such as voice/video recordings or perhaps all messages with a specific system keyword in the message XMPP stanza payload? Your app could then ensure to add that keyword so that only those messages are going to be retrieved.

**Files and Data API** - you could also use alternative options, for example use our Files API to upload files and our Data API to create a custom data structure which links Rooms and Media files. 

I think it makes sense at this stage to have a quick call perhaps so we can discuss the best implementation options together. Feel free to use my Calendly link if this works for you: https://calendly.com/taras-f

**Comment by seshubonam** [comment]
sure ill connect over a call on 18th. whats the best slot for you in UK time.

**Comment by phwizard** [comment]
@seshubonam tomorrow is flexible. anything between 10am - 6pm UK should work

**Comment by seshubonam** [comment]
scheduled 11am. thanks.

**Comment by phwizard** [comment]
@seshubonam apologies just realized I forgot to add one meeting into Calendar and it's exactly 11am 

do you mind moving our call with you to anywhere around 8am, 9am, 10am or 1pm UK onwards?

**Comment by seshubonam** [comment]
sure no issue with the time.

**Comment by phwizard** [comment]
@seshubonam is 10am UK good for you?

**Comment by seshubonam** [comment]
yes works

**Comment by seshubonam** [comment]
yes works

**Comment by seshubonam** [comment]
cal only shows10:30 so bokked that. I will be available at 10 anyway.

**Comment by seshubonam** [comment]
ok got your updated time. all good. 10am, thanks

**Comment by seshubonam** [comment]
ok got your updated time. all good. 10am, thanks

**Comment by phwizard** [comment]
OK sure, hope all is well and please use Calendly link for the next week

**Comment by dzinzyura** [comment]
Hi @seshubonam!

I hope this message finds you well. 

I wanted to touch base regarding the call. Is it still necessary at this point? If yes, please feel free to choose a more convenient time using the Calendly link:

https://calendly.com/taras-f

Let's make sure our conversation fits seamlessly into your schedule. Looking forward to your response!

Best regards,
Vladyslav
Ethora team

**Comment by seshubonam** [comment]
yes the call is a priority. Have not been well and some development was going to better demo the app. I will book a slot for wednesday. thanks for your continued support.

**Comment by phwizard** [comment]
@seshubonam please note we have added description of chat-related client-server flow here: https://github.com/dappros/ethora/blob/main/api/chats.md

For example, for your purposes of getting X amount of messages per room you can use the getRoomArchiveStanza function which uses the XMPP mechanism. 

In any case, let's discuss your requirements on our call today and we'll be able to advise better.

**Comment by seshubonam** [comment]
sure thanks. looking forward to the call.

**Comment by phwizard** [comment]
Great to talk Seshu and thank you for your efforts in making the product more known, definitely that should help the ecosystem! As discussed, we will be happy to help from our side with anything, just let us know.

**Comment by phwizard** [comment]
closing this for now, please reopen if any new questions
